File: src/content/docs/topics/cs/design-patterns/prototype.ts
// Prototype interface with a covariant return type
interface Cloneable {
  clone(): this;
}

// --- Concrete types ---

class TextDocument implements Cloneable {
  constructor(
    public title: string,
    public body: string,
    public tags: string[],
  ) {}

  clone(): this {
    // Shallow-copy the object, then deep-copy mutable fields
    const copy = Object.create(Object.getPrototypeOf(this)) as this;
    Object.assign(copy, this);
    copy.tags = [...this.tags]; // tags array is mutable, copy it
    return copy;
  }

  toString(): string {
    return `TextDocument("${this.title}", tags=[${this.tags.join(", ")}])`;
  }
}

class SpreadsheetDocument implements Cloneable {
  constructor(
    public title: string,
    public rows: number[][],
  ) {}

  clone(): this {
    const copy = Object.create(Object.getPrototypeOf(this)) as this;
    Object.assign(copy, this);
    copy.rows = this.rows.map((row) => [...row]); // deep-copy 2D array
    return copy;
  }

  toString(): string {
    return `SpreadsheetDocument("${this.title}", ${this.rows.length} rows)`;
  }
}

// --- Registry ---

class DocumentRegistry {
  private items = new Map<string, Cloneable>();

  register(key: string, prototype: Cloneable): void {
    this.items.set(key, prototype);
  }

  get<T extends Cloneable>(key: string): T {
    const proto = this.items.get(key);
    if (!proto) throw new Error(`No prototype registered for key "${key}"`);
    return proto.clone() as T;
  }
}

// --- Usage ---

const registry = new DocumentRegistry();

registry.register(
  "report",
  new TextDocument("Quarterly Report", "Summary goes here.", ["finance", "q1"]),
);

registry.register(
  "budget",
  new SpreadsheetDocument("Budget Template", [[0, 0, 0]]),
);

const doc1 = registry.get<TextDocument>("report");
doc1.title = "Q2 Report";
doc1.tags.push("q2");

const doc2 = registry.get<TextDocument>("report");
// doc2 still has the original title and tags -- the registry prototype is untouched

console.log(doc1.toString()); // TextDocument("Q2 Report", tags=[finance, q1, q2])
console.log(doc2.toString()); // TextDocument("Quarterly Report", tags=[finance, q1])

File: src/content/docs/topics/cs/design-patterns/prototype.go
package main

import "fmt"

// Go has no built-in clone mechanism.
// The convention is a Clone() method on each concrete type.
// Watch for shallow vs. deep copy: pointer fields and slices share backing
// memory unless you explicitly copy them.

// --- Prototype interface ---

type Document interface {
	Clone() Document
	Title() string
}

// --- TextDocument ---

type TextDocument struct {
	title string
	body  string
	tags  []string // slice header is copied by value, but backing array is shared
}

func NewTextDocument(title, body string, tags []string) *TextDocument {
	return &TextDocument{title: title, body: body, tags: tags}
}

// Clone performs a deep copy of the mutable slice field.
func (d *TextDocument) Clone() Document {
	tagsCopy := make([]string, len(d.tags))
	copy(tagsCopy, d.tags)
	return &TextDocument{
		title: d.title,
		body:  d.body,
		tags:  tagsCopy,
	}
}

func (d *TextDocument) Title() string { return d.title }

func (d *TextDocument) AddTag(tag string) { d.tags = append(d.tags, tag) }

func (d *TextDocument) String() string {
	return fmt.Sprintf("TextDocument(%q, tags=%v)", d.title, d.tags)
}

// --- SpreadsheetDocument ---

type SpreadsheetDocument struct {
	title string
	rows  [][]int // 2D slice: outer slice AND inner slices must be copied
}

func NewSpreadsheetDocument(title string, rows [][]int) *SpreadsheetDocument {
	return &SpreadsheetDocument{title: title, rows: rows}
}

// Clone deep-copies the 2D slice.
func (s *SpreadsheetDocument) Clone() Document {
	rowsCopy := make([][]int, len(s.rows))
	for i, row := range s.rows {
		rowCopy := make([]int, len(row))
		copy(rowCopy, row)
		rowsCopy[i] = rowCopy
	}
	return &SpreadsheetDocument{title: s.title, rows: rowsCopy}
}

func (s *SpreadsheetDocument) Title() string { return s.title }

func (s *SpreadsheetDocument) String() string {
	return fmt.Sprintf("SpreadsheetDocument(%q, %d rows)", s.title, len(s.rows))
}

// --- Registry ---

type DocumentRegistry struct {
	items map[string]Document
}

func NewDocumentRegistry() *DocumentRegistry {
	return &DocumentRegistry{items: make(map[string]Document)}
}

func (r *DocumentRegistry) Register(key string, proto Document) {
	r.items[key] = proto
}

func (r *DocumentRegistry) Get(key string) (Document, error) {
	proto, ok := r.items[key]
	if !ok {
		return nil, fmt.Errorf("no prototype registered for key %q", key)
	}
	return proto.Clone(), nil
}

// --- Main ---

func main() {
	registry := NewDocumentRegistry()

	registry.Register("report", NewTextDocument(
		"Quarterly Report", "Summary goes here.", []string{"finance", "q1"},
	))

	doc1, _ := registry.Get("report")
	td1 := doc1.(*TextDocument)
	td1.title = "Q2 Report"
	td1.AddTag("q2")

	doc2, _ := registry.Get("report")
	td2 := doc2.(*TextDocument)

	fmt.Println(td1) // TextDocument("Q2 Report", tags=[finance q1 q2])
	fmt.Println(td2) // TextDocument("Quarterly Report", tags=[finance q1])

	// Shallow copy pitfall (do NOT do this):
	// bad := *td1   // copies slice header, not backing array
	// bad.tags[0] = "OOPS"  // would mutate td1.tags too
}
